docs/architecture/adr-105-refactor-mempool-senders.md

```diff
@@ -0,0 +1,141 @@
+mempool's `txs` data structure. However, the list of senders is an important
+component of the propagation protocol and it should be part of the reactor,
+while `txs` is part of the implementation of this specific version (v0) of the
+mempool.
+
+In this document, we propose a solution that involves moving the list of senders
+from the mempool implementation to the reactor. This change will simplify the
+code, establish a more clear separation of the propagation protocol and the data
+structure, and allow for future improvements to the mempool as a whole.
+
+## Detailed Design
+ 
+We propose the following changes to the mempool's reactor and data structure.
+
+- In the `Mempool` interface, change the signature of `CheckTx` from
+    ``` golang
+    CheckTx(tx types.Tx, cb func(*abci.ResponseCheckTx), txInfo TxInfo) error
+    ```
+    to
+    ``` golang
+    CheckTx(tx types.Tx) (abcicli.ReqRes, error)
+    ```
+  - The returning `ReqRes` object can be used to set and invoke a callback to
+    handle the response, if needed.
+  - The callback parameter `cb` is no longer needed. Currently, this is mainly
+    used by the RPC endpoints `broadcast_tx_sync` and `broadcast_tx_commit`, and
+    in tests for checking that the response is valid. However, the same
+    functionality can be obtained with `ReqRes` response.
+  - `txInfo` contains information about the sender, which is also no longer
+    needed, as justified by the next point.
+- The list of senders for each transaction is currently stored in `mempoolTx`,
+  the data structure for the entries of `txs`. Move the senders out of
+  `mempoolTx` to a new map `txSenders` of type `map[types.TxKey]map[uint16]bool`
+  in the mempool reactor. `txSenders` would map transaction keys to a set of
+  peer ids (of type `uint16`). Add also a `cmtsync.RWMutex` lock to handle
+  concurrent accesses to the map.
+  - This refactoring should not change the fact that the list of senders live as
+    long as the transactions are in the mempool. When a transaction is received
+    by the reactor (either via RPC or P2P), we call `CheckTx`. We know whether a
+    transaction is valid and was included in the mempool by reading the `ReqRes`
+    response. If this is the case, record the list of senders in `txSenders`.
+    When a transaction is removed from the mempool, notify the reactor to remove
+    the list of senders for that transaction, with the channel described below.
+- In `CListMempool`, introduce a new channel `txsRemoved` of type `chan
+  types.TxKey` to notify the mempool reactor that a transaction was removed from
+  the mempool.
+  - This is the same mechanism used to notify the consensus reactor when there
+  are transactions available to include in a block.
+- In the mempool reactor, introduce a new goroutine to handle incoming
+  transaction keys from the `txsRemoved` channel. For each key received, update
+  `txSenders`.
```

Comment:
Are we sure there's no possibility here (even if we're using locks) that the info flowing on the `txsRemoved` channel "conflicts" with info flowing in the other direction? What I'm thinking of (not 100% clear in my mind) is the following.
More or less at the same time:
* A TX (tx1) is removed and placed in the channel
* A newly received TX (tx2) passes `CheckTx` and is added to the mempool
* tx1 and tx2 have the same sender

I don't know, I have the impression that, rather than a channel and a goroutine, you want a callback set by the reactor, and called by the mempool implementation. But I'm not 100% sure.

Reply:
Each transaction has its own list of sender IDs. So in your scenario there are no conflicts: tx1 will lose a sender, and tx2 will gain one, even when they are the sender. 

I didn't considered a callback instead of a channel and a goroutine. Probably it has its benefits or it's more idiomatic, but I'm not 100% sure either.

Reply:
I added the callback option as an alternative in the ADR. After playing with both implementations, now I think a callback will be the better option.